src/main/scala/viper/gobra/translator/encodings/DomainEncoding.scala

```diff
@@ -0,0 +1,101 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.translator.encodings
+
+import org.bitbucket.inkytonik.kiama.==>
+import viper.gobra.ast.{internal => in}
+import viper.gobra.theory.Addressability.{Exclusive, Shared}
+import viper.gobra.translator.Names
+import viper.gobra.translator.interfaces.Context
+import viper.gobra.translator.util.ViperWriter.{CodeWriter, MemberWriter}
+import viper.silver.{ast => vpr}
+
+class DomainEncoding extends LeafTypeEncoding {
+
+  import viper.gobra.translator.util.ViperWriter.CodeLevel._
+  import viper.gobra.translator.util.ViperWriter.{MemberLevel => ml}
+  import viper.gobra.translator.util.TypePatterns._
+
+  /**
+    * Translates a type into a Viper type.
+    */
+  override def typ(ctx: Context): in.Type ==> vpr.Type = {
+    case ctx.Domain(domain) / m =>
+      m match {
+        case Exclusive => domainType(domain.name)
+        case Shared    => vpr.Ref
+      }
+  }
+
+  private def domainType(domainName: String): vpr.DomainType = vpr.DomainType(domainName, Map.empty)(Seq.empty)
+
+  override def member(ctx: Context): in.Member ==> MemberWriter[Vector[vpr.Member]] = {
+    case domain: in.DomainDefinition =>
+      val domainName = domain.name
+      val (dPos, dInfo, dErrT) = domain.vprMeta
+
+      val funcs = domain.funcs map { f =>
+        val (fPos, fInfo, fErrT) = f.vprMeta
+        val formalArgs = f.args map ctx.typeEncoding.variable(ctx)
+        val resType = ctx.typeEncoding.typ(ctx)(f.results.typ)
+        vpr.DomainFunc(f.name.name, formalArgs, resType)(fPos, fInfo, domainName, fErrT)
+      }
+
+      val defaultFunc = vpr.DomainFunc(
```

Comment:
It seems a bit weird to me that there is always a default function generated under the hood for every domain. Do we need this? Does this lead to weird side effects that one cannot state disequalities between default function and a user-defined constructor?

Reply:
In the current infrastructure, everything must have a default value. I do not know what you mean with "disequalities between default function and a user-defined constructor".